File: feb-08-longest-balanced-subarray-ii/solution.java
import java.util.Arrays;

class Solution {
    private int[] mn;
    private int[] mx;
    private int[] lazy;

    private void apply(int idx, int delta) {
        mn[idx] += delta;
        mx[idx] += delta;
        lazy[idx] += delta;
    }

    private void push(int idx) {
        int delta = lazy[idx];
        if (delta == 0) {
            return;
        }
        int left = idx << 1;
        int right = left | 1;

        mn[left] += delta;
        mx[left] += delta;
        lazy[left] += delta;

        mn[right] += delta;
        mx[right] += delta;
        lazy[right] += delta;

        lazy[idx] = 0;
    }

    private void pull(int idx) {
        int left = idx << 1;
        int right = left | 1;
        mn[idx] = Math.min(mn[left], mn[right]);
        mx[idx] = Math.max(mx[left], mx[right]);
    }

    private void rangeAdd(int idx, int l, int r, int ql, int qr, int delta) {
        if (ql <= l && r <= qr) {
            apply(idx, delta);
            return;
        }

        push(idx);
        int mid = (l + r) >>> 1;
        if (ql <= mid) {
            rangeAdd(idx << 1, l, mid, ql, qr, delta);
        }
        if (qr > mid) {
            rangeAdd((idx << 1) | 1, mid + 1, r, ql, qr, delta);
        }
        pull(idx);
    }

    private int firstZero(int idx, int l, int r, int ql, int qr) {
        if (ql <= l && r <= qr) {
            if (mn[idx] > 0 || mx[idx] < 0) {
                return -1;
            }
            if (l == r) {
                return l;
            }
        } else if (r < ql || l > qr) {
            return -1;
        }

        if (l == r) {
            return -1;
        }

        push(idx);
        int mid = (l + r) >>> 1;
        if (ql <= mid) {
            int leftRes = firstZero(idx << 1, l, mid, ql, qr);
            if (leftRes != -1) {
                return leftRes;
            }
        }
        if (qr > mid) {
            return firstZero((idx << 1) | 1, mid + 1, r, ql, qr);
        }
        return -1;
    }

    public int longestBalanced(int[] nums) {
        int n = nums.length;
        int size = n * 4;
        mn = new int[size];
        mx = new int[size];
        lazy = new int[size];

        int[] lastPos = new int[100001];
        Arrays.fill(lastPos, -1);

        int ans = 0;
        for (int r = 0; r < n; r++) {
            int x = nums[r];
            int left = lastPos[x] + 1;
            int delta = (x & 1) == 0 ? 1 : -1;

            rangeAdd(1, 0, n - 1, left, r, delta);
            lastPos[x] = r;

            int l = firstZero(1, 0, n - 1, 0, r);
            if (l != -1) {
                ans = Math.max(ans, r - l + 1);
            }
        }

        return ans;
    }
}

File: feb-08-longest-balanced-subarray-ii/solution.py
from typing import List


class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        n = len(nums)
        seg_size = 4 * n
        mn = [0] * seg_size
        mx = [0] * seg_size
        lazy = [0] * seg_size

        def apply(i: int, delta: int) -> None:
            mn[i] += delta
            mx[i] += delta
            lazy[i] += delta

        def push(i: int) -> None:
            delta = lazy[i]
            if delta == 0:
                return
            left = i << 1
            right = left | 1
            mn[left] += delta
            mx[left] += delta
            lazy[left] += delta
            mn[right] += delta
            mx[right] += delta
            lazy[right] += delta
            lazy[i] = 0

        def pull(i: int) -> None:
            left = i << 1
            right = left | 1
            mn[i] = mn[left] if mn[left] < mn[right] else mn[right]
            mx[i] = mx[left] if mx[left] > mx[right] else mx[right]

        def range_add(i: int, l: int, r: int, ql: int, qr: int, delta: int) -> None:
            if ql <= l and r <= qr:
                apply(i, delta)
                return
            push(i)
            m = (l + r) >> 1
            if ql <= m:
                range_add(i << 1, l, m, ql, qr, delta)
            if qr > m:
                range_add((i << 1) | 1, m + 1, r, ql, qr, delta)
            pull(i)

        def first_zero(i: int, l: int, r: int, ql: int, qr: int) -> int:
            if ql <= l and r <= qr:
                if mn[i] > 0 or mx[i] < 0:
                    return -1
                if l == r:
                    return l
            elif r < ql or l > qr:
                return -1

            if l == r:
                return -1

            push(i)
            m = (l + r) >> 1
            if ql <= m:
                left_res = first_zero(i << 1, l, m, ql, qr)
                if left_res != -1:
                    return left_res
            if qr > m:
                return first_zero((i << 1) | 1, m + 1, r, ql, qr)
            return -1

        last_pos = [-1] * 100001
        ans = 0

        for r, x in enumerate(nums):
            left = last_pos[x] + 1
            delta = 1 if (x & 1) == 0 else -1
            range_add(1, 0, n - 1, left, r, delta)
            last_pos[x] = r

            l = first_zero(1, 0, n - 1, 0, r)
            if l != -1 and r - l + 1 > ans:
                ans = r - l + 1

        return ans
